### PR TITLE
Complete inside class constant declaration

### LINKED ISSUE #1978: Class completion does not work inside a constant value
image

### PR DESCRIPTION
**Pull request by przepompownia**
Closes #1978

### CODE REVIEW

**Comment by przepompownia** [comment]
I'm not sure if it's the correct logic but first tests suggest that it works.

File: lib/Completion/Tests/Integration/Bridge/TolerantParser/ReferenceFinder/ExpressionNameCompletorTest.php [hunk 1]
```diff
@@ -96,6 +96,16 @@ public function provideComplete(): Generator
                 ]
             ]
         ];
+
+        yield 'inside class constant declaration' => [
+            '<?php class Foobar { private const FOO; private const BAR = self::F<>  }', [
+                [
+                    'type'              => Suggestion::TYPE_CONSTANT,
+                    'name'              => 'self::FOO',
```
Comment (przepompownia) [inline lib/Completion/Tests/Integration/Bridge/TolerantParser/ReferenceFinder/ExpressionNameCompletorTest.php:104]: Please check if the expectation is correct.

File: lib/Completion/Bridge/TolerantParser/CompletionContext.php [hunk 2]
```diff
@@ -42,6 +43,10 @@ public static function expression(?Node $node): bool
             return false;
         }
 
+        if ($parent instanceof ConstElement) {
```
Comment (dantleech) [inline lib/Completion/Bridge/TolerantParser/CompletionContext.php:46]: think we can move this to the similar expressions in `return` below?

Comment (dantleech) [reply lib/Completion/Bridge/TolerantParser/CompletionContext.php:46]: or is it triggered by classMembersBody somehow?

File: lib/Completion/Tests/Integration/Bridge/TolerantParser/ReferenceFinder/ExpressionNameCompletorTest.php (hunk 1 above)
Comment (dantleech) [reply lib/Completion/Tests/Integration/Bridge/TolerantParser/ReferenceFinder/ExpressionNameCompletorTest.php:104]: why do you suspect it _not_ to be correct?

Comment (przepompownia) [reply lib/Completion/Tests/Integration/Bridge/TolerantParser/ReferenceFinder/ExpressionNameCompletorTest.php:104]: At the first I started with `name => 'FOO'` - the whole `'self::FOO'` caught my attention.

Comment (dantleech) [reply lib/Completion/Tests/Integration/Bridge/TolerantParser/ReferenceFinder/ExpressionNameCompletorTest.php:104]: hmm, yeah that doesn't seem quite right, did you test it in the editor?

Comment (przepompownia) [reply lib/Completion/Tests/Integration/Bridge/TolerantParser/ReferenceFinder/ExpressionNameCompletorTest.php:104]: Yes,

https://user-images.githubusercontent.com/11404453/206718275-a8e0bc8d-6d79-4524-80ba-277634802213.mp4

but I found that the new test case checks something that works already on master :grin: so I need to add the proper case before refactoring discussed in the other comment.

**Comment by przepompownia** [comment]
It's still not what I need:
image

**Comment by dantleech** [comment]
What are you missing ?

**Comment by dantleech** [comment]
it's a mine field, good luck!

**Comment by przepompownia** [comment]
https://user-images.githubusercontent.com/11404453/206793817-7a3d8915-d842-46bc-ac12-c3c973875c5a.mp4

Now it's time to return to the question where `$parent instanceof ConstElement` should be placed. Including this condition in `classMembersBody` requires repeat the query again in `expression()`.

**Comment by dantleech** [comment]
I think it's fine, its not a "query" but just an `instanceof` and it's true, it's not a direct child of the class members _bod but we should complete expressions on ConstElements.

I've just spent the past 30m trying to stop `function` and `const` being completed:

image

but I think we can merge this in it's current state as an improvement?

**Comment by przepompownia** [comment]
I agree that at the moment almost accurate suggestion list is better than none and we could merge this PR. 

Independently it would be good to keep in `master` PHPUnit test cases that fails (but should/could pass, like that shown by you on the picture below) but marked to exclude them with `--exclude-group` by CI. What do you think about such approach?

File: lib/Completion/Tests/Integration/Bridge/TolerantParser/ReferenceFinder/ExpressionNameCompletorTest.php (hunk 1 above)
Comment (dantleech) [reply lib/Completion/Tests/Integration/Bridge/TolerantParser/ReferenceFinder/ExpressionNameCompletorTest.php:104]: it is an artificial suggestion:

image

**Review by dantleech — APPROVED** [review]

**Comment by dantleech** [comment]
thanks,I updated the changelog and changed the test to simulate something more realistic...

**Pull request merged**